README.md
# postcss-speech-bubble
postcss-speech-bubble creates speech bubbles in just a couple of lines of CSS.

## Installation
npm install postcss-speech-bubble

## Usage
postcss-speech-bubble offers three rules that can be used to build different kinds of speech bubbles.

bubble: borderSize borderRadius type color;
* borderSize: border size in px
* borderRadius: border radius on the speech bubble
* type:
    solid or hollow
    Solid creates a speech bubble that uses the color provided here
    as background color.
    Hollow uses the color provided in this rule as the border color.
* color:
    Background color of the bubble and the beaker if it is solid
    border color on bubble and beaker if it is hollow

bubble-beaker: beakerSize positionOfBeaker;
* beakerSize: size of the speech beaker. Please provide this in px.
* positionOfBeaker: Where the beaker should be for the speech bubble. Below are possiblePosition's options:
  * top-right
  * top-left
  * top-center
  * bottom-right
  * bottom-left
  * bottom-center
  * left-top
  * left-bottom
  * left-middle
  * right-top
  * right-bottom
  * right-middle

bubble-background: color;
This is necessary if you need to provide a bubble with a border and a background color.
You can define these bubbles by making them hollow and providing the border color and providing a background color through this property.

#### Important Note: Please provide width and height for these bubbles or let the content specify the width/height. The bubbles will not look right without proper width/height. 

## Examples

### Solid bubble (No border)
![solid bubble](https://lh3.googleusercontent.com/-AeIItjhWS2c/VqRriS6DYoI/AAAAAAAAOzw/JSMxzDnBag4/s338-Ic42/Screen%252520Shot%2525202016-01-23%252520at%25252010.07.57%252520PM.png)

    .bubble {
      bubble-beaker: 12px top-right;
      bubble: 0 0 solid lightGrey;
      width: 140px;
      height: 80px;
    }

### Hollow bubble
![hollow bubble](https://lh3.googleusercontent.com/-kvMLnldOwk4/VqRrhyoT5xI/AAAAAAAAOzs/nElrVn57kZE/s386-Ic42/Screen%252520Shot%2525202016-01-23%252520at%25252010.09.20%252520PM.png)

    .bubble {
      bubble: 1px 10px hollow black;
      bubble-beaker: 10px left-middle;
      width: 150px;
      height: 100px;
    }

### Hollow bubble with a background
![hollow bubble with background](https://lh3.googleusercontent.com/-_WS8rmal0Vs/VqRrhbE9lJI/AAAAAAAAOzk/Z2eg19MSDzA/s282-Ic42/Screen%252520Shot%2525202016-01-23%252520at%25252010.12.48%252520PM.png)

    .bubble {
      bubble-beaker: 12px right-middle;
      bubble: 3px 0 hollow black;
      bubble-background: #E44146;
      width: 100px;
      height: 120px;
    }

## [Changelog] (./CHANGELOG.md "Changelog")
## [License](./LICENSE "License")


CHANGELOG.md
#1.0.12
Added an important note to readme about specifying width and height

#1.0.1, #1.0.11 - 2016-01-24
Fixed the Git user name which was incorrect before.

# 0.0.9, 1.0.0 - 2016-01-23
First version of Speech bubble


test.js
import postcss from 'postcss';
import test    from 'ava';
import fs from 'fs';
import bubble from './';

require.extensions['.css'] = function (module, filename) {
  module.exports = fs.readFileSync(filename, 'utf8');
};

var inputCSS = require('./test/input.css');
var outputCSS = require('./test/output.css');

var inputBubbleNotCorrect = require('./test/input_bubbleIncorrect.css');
var inputBeakerNotSpecified = require('./test/input_bubbleBeakerNotSpecified.css');
var inputBeakerNotSpecifiedCorrectly = require('./test/input_beakerIncorrect.css');
var inputBeakerPositionDirectionIncorrect = require('./test/input_beakerPositionDirectionIncorrect.css');

function run(t, input, output, opts = { }) {
    return postcss([ bubble(opts) ]).process(input)
        .then( result => {
            if (opts.shouldThrowError) {
              t.fail();
            }
            t.same(result.css, output);
            t.same(result.warnings().length, opts.error || 0);
        })
        .catch( () => {
          if (opts.shouldThrowError) {
            t.pass();
          }
        });
}

test('converts bubble values to CSS', t => {
    return run(t, inputCSS.toString(), outputCSS.toString(), { });
});

test('throws exception when bubble is not specified correctly', t => {
    return run(t, inputBubbleNotCorrect, '', { shouldThrowError: true });
});

test('throws exception when bubble-beaker is not specified', t => {
    return run(t, inputBeakerNotSpecified, '', { shouldThrowError: true });
});

test('throws exception when bubble-beaker is not specified correctly', t => {
    return run(t, inputBeakerNotSpecifiedCorrectly, '',
      { shouldThrowError: true });
});

test('throws exception when bubble-beaker position/direction is not specified correctly', t => {
    return run(t, inputBeakerPositionDirectionIncorrect, '',
      { shouldThrowError: true });
});


index_dev.js
const postcss = require('postcss');

const getBeakerDeclarationValues = (rule) => {
  let bubbleBeakerValues = [];
  let position = null;
  let bubbleBackColor = null;

  for (let node of rule.nodes) {
    switch (node.prop.trim().toLowerCase()) {
      case 'bubble-beaker':
        bubbleBeakerValues = node.value.split(' ');
        rule.removeChild(node);
        break;
      case 'position':
        position = node.value;
        break;
      case 'bubble-background':
        bubbleBackColor = node.value;
        rule.removeChild(node);
        break;
    }
  }
  return {
    bubbleBeakerValues,
    position,
    bubbleBackColor
  };
};

const extractBubbleAndBeakerSpecs = (bubbleValues, beakerValues) => {
  let bubble = {};
  let beaker = {};

  if (bubbleValues && bubbleValues.length !== 4) {
    throw `Bubble is not correctly specified. Please check the docs
    and provide the right values.`;
  }

  if (beakerValues && beakerValues.length !== 2) {
    throw `Bubble beaker is not specified correctly. Please check the docs and
    provide the right values.`;
  }

  [bubble.borderSize, bubble.borderRadius, bubble.type, bubble.color] = bubbleValues;

  if (beakerValues[1].indexOf('-') >= 0) {
    beaker.direction = beakerValues ? (beakerValues[1].split('-'))[0] : null;
    beaker.position = beakerValues ? (beakerValues[1].split('-'))[1] : null;
  } else {
    throw `Beaker is not provided correctly. Please check the docs and
    provide the right values.`;
  }

  let possibleDirections = ['top', 'bottom', 'left', 'right'];
  let possiblePositions = ['top', 'bottom', 'left', 'right', 'center', 'middle'];

  if (possibleDirections.indexOf(beaker.direction) < 0 ||
    possiblePositions.indexOf(beaker.position) < 0) {
    throw `Beaker is not provided correctly. Please specify bubble-beaker:
      <beaker size> <beaker position>. Beaker position should be top-left,
      top-right, bottom-center, left-middle, etc.`;
  }

  beaker.size = beakerValues ? beakerValues[0] : null;

  if (beaker.size.indexOf('px') < 0) {
    throw 'Expecting a px value for beaker size. Please do not provide other units';
  }

  return {
    bubble,
    beaker
  };
};

const addBubble = (bubble, beaker, position, rule) => {
  // Add border size for the bubble
  if (parseInt(bubble.borderSize, 10) > 0) {
    rule.append('border: ' + bubble.borderSize + ' solid ' + bubble.color);
  }

  if (bubble.type.toLowerCase() === 'solid') {
    rule.append('background-color: ' + bubble.color);
  }

  if (bubble.backColor) {
    rule.append('background-color: ' + bubble.backColor);
  }

  // Add border radius if any
  if (parseInt(bubble.borderRadius, 10) > 0) {
    rule.append('border-radius: ' + bubble.borderRadius);
  }

  // Adjust the positioning of the bubble based on beaker
  if (beaker.size && beaker.direction
    && parseInt(beaker.size, 10) > 0) {
    rule.append('margin-' + beaker.direction + ':' + beaker.size);
  }

  // Add position relative to container so the beaker will appear correctly.
  if (!position) {
    rule.append('position: relative');
  } else if (position !== 'absolute' || position !== 'relative') {
    throw 'Position for the bubble should either be absolute or relative.';
  }
};

const addBeaker = (bubble, beaker, rule) => {
  // Add the beaker now
  let beforeRule = postcss.rule({selector: rule.selector + ':before'});
  let afterRule = null;

  if (bubble.type === 'hollow') {
    // We need to add another beaker the same background color as the bubble
    // to create a caret look instead of triangle.
    afterRule = postcss.rule({selector: rule.selector + ':after'});
  }

  var commonRules = [
    'content: \'\'',
    'position: absolute'
  ];

  const afterColor = bubble.backColor ? bubble.backColor : 'white';
  const transparentTriangle = beaker.size + ' solid transparent';
  const solidTriangle = beaker.size + ' solid ' + bubble.color;
  let border = parseInt(bubble.borderSize.split('px')[0], 10) || 1;
  border = border > 1 ? border + 1 : 1;


  // First add solid colored beaker
  switch(beaker.direction.toLowerCase()) {
    case 'top':
      beforeRule.append('top: -' + beaker.size);
      commonRules.push('border-left: ' + transparentTriangle);
      commonRules.push('border-right: ' + transparentTriangle);
      beforeRule.append('border-bottom: ' + solidTriangle);
      if (afterRule) {
        afterRule.append('top: -' + (parseInt(beaker.size.split('px')[0], 10) - border) + 'px');
        afterRule.append('border-bottom: ' + beaker.size + ' solid ' + afterColor);
      }
      break;
    case 'bottom':
      beforeRule.append('bottom: -' + beaker.size);
      commonRules.push('border-left: ' + transparentTriangle);
      commonRules.push('border-right: ' + transparentTriangle);
      beforeRule.append('border-top: ' + solidTriangle);
      if (afterRule) {
        afterRule.append('bottom: -' + (parseInt(beaker.size.split('px')[0], 10) - border) + 'px');
        afterRule.append('border-top: ' + beaker.size + ' solid ' + afterColor);
      }
      break;
    case 'left':
      beforeRule.append('left: -' + beaker.size);
      commonRules.push('border-top: ' + transparentTriangle);
      commonRules.push('border-bottom: ' + transparentTriangle);
      beforeRule.append('border-right: ' + solidTriangle);
      if (afterRule) {
        afterRule.append('left: -' + (parseInt(beaker.size.split('px')[0], 10) - border) + 'px');
        afterRule.append('border-right: ' + beaker.size + ' solid ' + afterColor);
      }
      break;
    case 'right':
      beforeRule.append('right: -' + beaker.size);
      commonRules.push('border-top: ' + transparentTriangle);
      commonRules.push('border-bottom: ' + transparentTriangle);
      beforeRule.append('border-left: ' + solidTriangle);
      if (afterRule) {
        afterRule.append('right: -' + (parseInt(beaker.size.split('px')[0], 10) - border) + 'px');
        afterRule.append('border-left: ' + beaker.size + ' solid ' + afterColor);
      }
      break;
    default:
      throw `Provide correct position for bubble-beaker. Some examples
        are top-right, top-center, left-middle, left-top`;
  }

  const beakerPos = parseInt(beaker.size.split('px')[0], 10) * 1.5 + 'px';
  switch(beaker.position.toLowerCase()) {
    case 'left':
      commonRules.push('left: ' + beakerPos);
      break;
    case 'right':
      commonRules.push('right: ' + beakerPos);
      break;
    case 'center':
      commonRules.push('left: calc(50% - ' + beaker.size + ')');
      commonRules.push('transformX: -50%');
      break;
    case 'middle':
      commonRules.push('top: calc(50% - ' + beaker.size + ')');
      commonRules.push('transformY: -50%');
      break;
    case 'top':
      commonRules.push('top: ' + beakerPos);
      break;
    case 'bottom':
      commonRules.push('bottom: ' + beakerPos);
      break;
  }

  for (let rule of commonRules) {
    beforeRule.append(rule);
    if(afterRule) afterRule.append(rule);
  }

  rule.parent.nodes.push(beforeRule);
  if (afterRule) rule.parent.nodes.push(afterRule);
};

module.exports = postcss.plugin('postcss-bubble', function (opts) {
    opts = opts || {};

    return function (css) {
      css.walkDecls(function (decl, i) {
        let rule = decl.parent;
        let value = decl.prop;

        if (decl.prop === 'bubble') {
          // Get bubble beaker if specified
          let bubble = {};
          let beaker = {};

          const obj = getBeakerDeclarationValues(rule);
          const bubbleBeakerValues = obj.bubbleBeakerValues;
          const position = obj.position;
          const bubbleBackColor = obj.bubbleBackColor;

          // Consolidate all the values to construct
          const bubbleValues = decl.value.split(' ');

          const specs = extractBubbleAndBeakerSpecs(bubbleValues, bubbleBeakerValues);
          bubble = specs.bubble;
          bubble.backColor = bubbleBackColor;
          beaker = specs.beaker;

          addBubble(bubble, beaker, position, rule);
          addBeaker(bubble, beaker, rule);

          rule.removeChild(decl);
        }

      });
    };
});


index.js
'use strict';

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

var postcss = require('postcss');

var getBeakerDeclarationValues = function getBeakerDeclarationValues(rule) {
  var bubbleBeakerValues = [];
  var position = null;
  var bubbleBackColor = null;

  var _iteratorNormalCompletion = true;
  var _didIteratorError = false;
  var _iteratorError = undefined;

  try {
    for (var _iterator = rule.nodes[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
      var node = _step.value;

      switch (node.prop.trim().toLowerCase()) {
        case 'bubble-beaker':
          bubbleBeakerValues = node.value.split(' ');
          rule.removeChild(node);
          break;
        case 'position':
          position = node.value;
          break;
        case 'bubble-background':
          bubbleBackColor = node.value;
          rule.removeChild(node);
          break;
      }
    }
  } catch (err) {
    _didIteratorError = true;
    _iteratorError = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion && _iterator.return) {
        _iterator.return();
      }
    } finally {
      if (_didIteratorError) {
        throw _iteratorError;
      }
    }
  }

  return {
    bubbleBeakerValues: bubbleBeakerValues,
    position: position,
    bubbleBackColor: bubbleBackColor
  };
};

var extractBubbleAndBeakerSpecs = function extractBubbleAndBeakerSpecs(bubbleValues, beakerValues) {
  var bubble = {};
  var beaker = {};

  if (bubbleValues && bubbleValues.length !== 4) {
    throw 'Bubble is not correctly specified. Please check the docs\n    and provide the right values.';
  }

  if (beakerValues && beakerValues.length !== 2) {
    throw 'Bubble beaker is not specified correctly. Please check the docs and\n    provide the right values.';
  }

  var _bubbleValues = _slicedToArray(bubbleValues, 4);

  bubble.borderSize = _bubbleValues[0];
  bubble.borderRadius = _bubbleValues[1];
  bubble.type = _bubbleValues[2];
  bubble.color = _bubbleValues[3];

  if (beakerValues[1].indexOf('-') >= 0) {
    beaker.direction = beakerValues ? beakerValues[1].split('-')[0] : null;
    beaker.position = beakerValues ? beakerValues[1].split('-')[1] : null;
  } else {
    throw 'Beaker is not provided correctly. Please check the docs and\n    provide the right values.';
  }

  var possibleDirections = ['top', 'bottom', 'left', 'right'];
  var possiblePositions = ['top', 'bottom', 'left', 'right', 'center', 'middle'];

  if (possibleDirections.indexOf(beaker.direction) < 0 || possiblePositions.indexOf(beaker.position) < 0) {
    throw 'Beaker is not provided correctly. Please specify bubble-beaker:\n      <beaker size> <beaker position>. Beaker position should be top-left,\n      top-right, bottom-center, left-middle, etc.';
  }

  beaker.size = beakerValues ? beakerValues[0] : null;

  if (beaker.size.indexOf('px') < 0) {
    throw 'Expecting a px value for beaker size. Please do not provide other units';
  }

  return {
    bubble: bubble,
    beaker: beaker
  };
};

var addBubble = function addBubble(bubble, beaker, position, rule) {
  // Add border size for the bubble
  if (parseInt(bubble.borderSize, 10) > 0) {
    rule.append('border: ' + bubble.borderSize + ' solid ' + bubble.color);
  }

  if (bubble.type.toLowerCase() === 'solid') {
    rule.append('background-color: ' + bubble.color);
  }

  if (bubble.backColor) {
    rule.append('background-color: ' + bubble.backColor);
  }

  // Add border radius if any
  if (parseInt(bubble.borderRadius, 10) > 0) {
    rule.append('border-radius: ' + bubble.borderRadius);
  }

  // Adjust the positioning of the bubble based on beaker
  if (beaker.size && beaker.direction && parseInt(beaker.size, 10) > 0) {
    rule.append('margin-' + beaker.direction + ':' + beaker.size);
  }

  // Add position relative to container so the beaker will appear correctly.
  if (!position) {
    rule.append('position: relative');
  } else if (position !== 'absolute' || position !== 'relative') {
    throw 'Position for the bubble should either be absolute or relative.';
  }
};

var addBeaker = function addBeaker(bubble, beaker, rule) {
  // Add the beaker now
  var beforeRule = postcss.rule({ selector: rule.selector + ':before' });
  var afterRule = null;

  if (bubble.type === 'hollow') {
    // We need to add another beaker the same background color as the bubble
    // to create a caret look instead of triangle.
    afterRule = postcss.rule({ selector: rule.selector + ':after' });
  }

  var commonRules = ['content: \'\'', 'position: absolute'];

  var afterColor = bubble.backColor ? bubble.backColor : 'white';
  var transparentTriangle = beaker.size + ' solid transparent';
  var solidTriangle = beaker.size + ' solid ' + bubble.color;
  var border = parseInt(bubble.borderSize.split('px')[0], 10) || 1;
  border = border > 1 ? border + 1 : 1;

  // First add solid colored beaker
  switch (beaker.direction.toLowerCase()) {
    case 'top':
      beforeRule.append('top: -' + beaker.size);
      commonRules.push('border-left: ' + transparentTriangle);
      commonRules.push('border-right: ' + transparentTriangle);
      beforeRule.append('border-bottom: ' + solidTriangle);
      if (afterRule) {
        afterRule.append('top: -' + (parseInt(beaker.size.split('px')[0], 10) - border) + 'px');
        afterRule.append('border-bottom: ' + beaker.size + ' solid ' + afterColor);
      }
      break;
    case 'bottom':
      beforeRule.append('bottom: -' + beaker.size);
      commonRules.push('border-left: ' + transparentTriangle);
      commonRules.push('border-right: ' + transparentTriangle);
      beforeRule.append('border-top: ' + solidTriangle);
      if (afterRule) {
        afterRule.append('bottom: -' + (parseInt(beaker.size.split('px')[0], 10) - border) + 'px');
        afterRule.append('border-top: ' + beaker.size + ' solid ' + afterColor);
      }
      break;
    case 'left':
      beforeRule.append('left: -' + beaker.size);
      commonRules.push('border-top: ' + transparentTriangle);
      commonRules.push('border-bottom: ' + transparentTriangle);
      beforeRule.append('border-right: ' + solidTriangle);
      if (afterRule) {
        afterRule.append('left: -' + (parseInt(beaker.size.split('px')[0], 10) - border) + 'px');
        afterRule.append('border-right: ' + beaker.size + ' solid ' + afterColor);
      }
      break;
    case 'right':
      beforeRule.append('right: -' + beaker.size);
      commonRules.push('border-top: ' + transparentTriangle);
      commonRules.push('border-bottom: ' + transparentTriangle);
      beforeRule.append('border-left: ' + solidTriangle);
      if (afterRule) {
        afterRule.append('right: -' + (parseInt(beaker.size.split('px')[0], 10) - border) + 'px');
        afterRule.append('border-left: ' + beaker.size + ' solid ' + afterColor);
      }
      break;
    default:
      throw 'Provide correct position for bubble-beaker. Some examples\n        are top-right, top-center, left-middle, left-top';
  }

  var beakerPos = parseInt(beaker.size.split('px')[0], 10) * 1.5 + 'px';
  switch (beaker.position.toLowerCase()) {
    case 'left':
      commonRules.push('left: ' + beakerPos);
      break;
    case 'right':
      commonRules.push('right: ' + beakerPos);
      break;
    case 'center':
      commonRules.push('left: calc(50% - ' + beaker.size + ')');
      commonRules.push('transformX: -50%');
      break;
    case 'middle':
      commonRules.push('top: calc(50% - ' + beaker.size + ')');
      commonRules.push('transformY: -50%');
      break;
    case 'top':
      commonRules.push('top: ' + beakerPos);
      break;
    case 'bottom':
      commonRules.push('bottom: ' + beakerPos);
      break;
  }

  var _iteratorNormalCompletion2 = true;
  var _didIteratorError2 = false;
  var _iteratorError2 = undefined;

  try {
    for (var _iterator2 = commonRules[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
      var _rule = _step2.value;

      beforeRule.append(_rule);
      if (afterRule) afterRule.append(_rule);
    }
  } catch (err) {
    _didIteratorError2 = true;
    _iteratorError2 = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion2 && _iterator2.return) {
        _iterator2.return();
      }
    } finally {
      if (_didIteratorError2) {
        throw _iteratorError2;
      }
    }
  }

  rule.parent.nodes.push(beforeRule);
  if (afterRule) rule.parent.nodes.push(afterRule);
};

module.exports = postcss.plugin('postcss-bubble', function (opts) {
  opts = opts || {};

  return function (css) {
    css.walkDecls(function (decl, i) {
      var rule = decl.parent;
      var value = decl.prop;

      if (decl.prop === 'bubble') {
        // Get bubble beaker if specified
        var bubble = {};
        var beaker = {};

        var obj = getBeakerDeclarationValues(rule);
        var bubbleBeakerValues = obj.bubbleBeakerValues;
        var position = obj.position;
        var bubbleBackColor = obj.bubbleBackColor;

        // Consolidate all the values to construct
        var bubbleValues = decl.value.split(' ');

        var specs = extractBubbleAndBeakerSpecs(bubbleValues, bubbleBeakerValues);
        bubble = specs.bubble;
        bubble.backColor = bubbleBackColor;
        beaker = specs.beaker;

        addBubble(bubble, beaker, position, rule);
        addBeaker(bubble, beaker, rule);

        rule.removeChild(decl);
      }
    });
  };
});
